### PR TITLE
use relative urls for redirects

### LINKED ISSUE #153: Flask/Werkzeug have moved to use relative URLs when redirecting
https://github.com/pallets/flask/blame/main/CHANGES.rst#L97

### PR DESCRIPTION
**Pull request by calebsyring**
fixes #153

### CODE REVIEW

File: keg_auth/model/__init__.py [hunk 1]
```diff
@@ -490,8 +490,8 @@ class BundleMixin(object):
     @might_commit
     @might_flush
     @classmethod
-    def edit(cls, oid=None, **kwargs):
```
Comment (guruofgentoo) [inline keg_auth/model/__init__.py:493]: What's the reason behind this change? It doesn't seem to be relevant to the issue, and the generic keg-elements method has `oid` (https://github.com/level12/keg-elements/blob/master/keg_elements/db/mixins.py#L195).

**Review by guruofgentoo — CHANGES_REQUESTED** [review]
The relevant updates look good to me. Not sure we want that model update in this revision, though.

File: keg_auth/model/__init__.py [hunk 2]
```diff
@@ -490,8 +490,8 @@ class BundleMixin(object):
...
-        obj = super(BundleMixin, cls).edit(oid=oid, _commit=False, **kwargs)
+    def edit(cls, id=None, **kwargs):
```
Comment (rsyring) [inline keg_auth/model/__init__.py:493]: nit: the reason I don't use "id" as a variable is because I don't want to shadow the id built-in function.

File: keg_auth/model/__init__.py (hunk 1 above)
Comment (calebsyring) [reply keg_auth/model/__init__.py:493]: This was a change I made to get the tests passing because `oid` was throwing back an "Unknown column or relationship" error. That made sense to me because I knew the column should be id, so I fixed it without investigating further, which I should not have done.

I've done a bit of digging, and it seems there's a small bug in KE. The `@kwargs_match_entity` decorator used on the `edit` method on the `MethodsMixin` (https://github.com/level12/keg-elements/blob/168e70af2b49bfd3f1e97705b86d158821e5d2ad/keg_elements/db/mixins.py#L193) will throw an error when `oid` is passed in as a kwarg (because it's not a column or relationship on the entity). We don't catch this in the tests because we pass the `oid` in without the keyword: https://github.com/level12/keg-elements/blob/168e70af2b49bfd3f1e97705b86d158821e5d2ad/keg_elements/tests/test_db_mixins.py#L77.

I created a pr in KE (https://github.com/level12/keg-elements/pull/175) and pushed up a commit to this pr with a better workaround.

Comment (rsyring) [reply keg_auth/model/__init__.py:493]: @calebsyring your last commit is the correct fix for this issue.

I believe this error started cropping up in tests due to a change @guruofgentoo made a few months ago to KE: https://github.com/level12/keg-elements/blame/168e70af2b49bfd3f1e97705b86d158821e5d2ad/keg_elements/db/mixins.py#L193

File: keg_auth/testing.py [hunk 3]
```diff
@@ -714,9 +714,9 @@ def test_next_parameter_not_open_redirect(self):
 
         assert resp.status_code == 302, resp.html
         # verify the 'next' parameter was ignored
-        assert resp.headers['Location'] == flask.url_for(
+        assert resp.headers['Location'] == urlparse(flask.url_for(
```
Comment (rsyring) [inline keg_auth/testing.py:717]: These tests are an anti-pattern.  We should be testing against a concrete expected value, not re-implementing the logic we use to calculate the URL in the main code.  E.g.:

```python
assert resp.headers['Location'] == '/login'
```

@guruofgentoo unless you object today, @calebsyring can make these adjustments while he's at it.

**Comment by rsyring** [comment]
The existing changes here also look good to me.  @calebsyring unless Matt objects, make the changes I requested to move the tests so they check a static value.  Then I think this PR will be GTG.

File: keg_auth/testing.py (hunk 3 above)
Comment (guruofgentoo) [reply keg_auth/testing.py:717]: Some injection either on the front side (mocking) or the tail end (as it is here), because this set of tests is meant to be run in any of our projects, which may have differing paths for each auth endpoint. I think putting that injection on the front side is cleaner, since it simplifies the assert.

Comment (rsyring) [reply keg_auth/testing.py:717]: But I think the point of these tests is to exercise the auth setup, so I don't think we'd want to mock the requests, would we?  Or, what would we be mocking?

Maybe what we should be doing here is having the app's override this base class and set hard-coded variables on the subclass to represent expected URLs.  Then, in this class, we'd do something like `assert resp.headers['Location'] == self.LOGIN_URL`.

```python
# in the app's test_auth.py
class TestAuthIntegration(AuthTests):
    user_ent = ents.User
    protected_url = '/users'
    protected_url_permissions = 'auth-manage'
   LOGIN_URL = '/login'
   FOO_URL = '/bar'
```

Thoughts?

Comment (guruofgentoo) [reply keg_auth/testing.py:717]: It'd essentially be a "config"-style mock on the auth manager to make the endpoint correspond to a known value.

Comment (calebsyring) [reply keg_auth/testing.py:717]: @guruofgentoo, how would that work? Since the tests are meant to be run in any of our projects, how could we guarantee the endpoint would exist without subclass attributes?

Comment (guruofgentoo) [reply keg_auth/testing.py:717]: On further thought, the above is correct - class attributes is the way to go here.
